edi_stock_oca/models/stock_picking.py

```diff
@@ -1,9 +1,41 @@
 # Copyright 2022 OCA
 # License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
 
+from contextlib import contextmanager
+
 from odoo import models
 
 
 class StockPicking(models.Model):
     _name = "stock.picking"
     _inherit = ["stock.picking", "edi.exchange.consumer.mixin"]
+
+    # Override all methods to trigger EDI exchange after picking state changes.
+    # ``stock.picking.state`` is computed so we cannot rely
+    # on the write as done by the consumer mixin.`
+    def action_confirm(self):
+        with self._edi_exchange_event_trigger():
+            result = super().action_confirm()
+        return result
+
+    def _action_done(self):
+        with self._edi_exchange_event_trigger():
+            result = super()._action_done()
+        return result
+
+    def action_cancel(self):
+        with self._edi_exchange_event_trigger():
+            result = super().action_cancel()
+        return result
+
+    @contextmanager
+    def _edi_exchange_event_trigger(self):
```

Comment:
Using this, I'd assume you want "before state change" notification to happen even if the action returns an error and doesn't change the state?

FWIW We faced a similar situation in another project with a stored computed field where we did override `_write` to trigger the notification only if the field name was in the values dict.

Reply:
No, if it fails here the transaction fails and nothing happens.
As for the `_write` I thought of it but I decided to keep it simple.
If we have at least another case falling in this situation we could override `_write` in the consumer mixin directly and check if `state` is computed.